codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java

```diff
@@ -25,32 +26,12 @@
 import java.util.Collections;
 import java.util.List;
 
+import static io.netty.util.internal.ObjectUtil.*;
+
 /**
  * Message container for decoded HAProxy proxy protocol parameters
  */
-public final class HAProxyMessage {
-
-    /**
-     * Version 1 proxy protocol message for 'UNKNOWN' proxied protocols. Per spec, when the proxied protocol is
-     * 'UNKNOWN' we must discard all other header values.
-     */
-    private static final HAProxyMessage V1_UNKNOWN_MSG = new HAProxyMessage(
-            HAProxyProtocolVersion.V1, HAProxyCommand.PROXY, HAProxyProxiedProtocol.UNKNOWN, null, null, 0, 0);
-
-    /**
-     * Version 2 proxy protocol message for 'UNKNOWN' proxied protocols. Per spec, when the proxied protocol is
-     * 'UNKNOWN' we must discard all other header values.
-     */
-    private static final HAProxyMessage V2_UNKNOWN_MSG = new HAProxyMessage(
-            HAProxyProtocolVersion.V2, HAProxyCommand.PROXY, HAProxyProxiedProtocol.UNKNOWN, null, null, 0, 0);
-
-    /**
-     * Version 2 proxy protocol message for local requests. Per spec, we should use an unspecified protocol and family
-     * for 'LOCAL' commands. Per spec, when the proxied protocol is 'UNKNOWN' we must discard all other header values.
-     */
-    private static final HAProxyMessage V2_LOCAL_MSG = new HAProxyMessage(
-            HAProxyProtocolVersion.V2, HAProxyCommand.LOCAL, HAProxyProxiedProtocol.UNKNOWN, null, null, 0, 0);
-
+public final class HAProxyMessage extends DefaultByteBufHolder {
```

Comment:
imho we should not extend `DefaultByteBufHolder` but `AbstractReferenceCounted`. There is no need to hold a reference to the "original" `ByteBuf` here.

Reply:
I like this idea, I think if `HAProxyMessage` inherits `AbstractReferenceCounted`, then maybe we can make the implementation a bit simpler. Just overwrite the `AbstractReferenceCounted.deallocate()` method and release the TLV list in this method, like this:
```java
class HAProxyMessage extends AbstractReferenceCounted {
    @Override
    public HAProxyMessage retain() {
        return (HAProxyMessage) super.retain();
    }
    // overwrite other methods...

    @Override
    protected void deallocate() {
       for (HAProxyTLV tlv : this.tlvs){
           tlv.release();
       }
    }
}
```
In this case, the TLV list will be released together when `HAProxyMessage` is released, but when `HAProxyMessage refCnt` is increased, the TLV `refCnt` will not increase with `HAProxyMessage`.

@normanmaurer What do you think of this? Will it cause other additional problems?

By the way, I found that this is also the case in DNSMessage:
https://github.com/netty/netty/blob/00afb19d7a37de21b35ce4f6cb3fa7f74809f2ab/codec-dns/src/main/java/io/netty/handler/codec/dns/AbstractDnsMessage.java#L378-L388

Reply:
@qeesung good question...Let me think about it but in general I would say it is fine.

Reply:
@normanmaurer It's done in this way. Could you please review it again :-)?